adding python all_gather coalesced functionality and testing.

agolynski opened the pull request:
Summary:
caveat 1: this only works in sync mode.
caveat 2: this is going to go away and be replaced by c++ implementation

Test Plan: buck test caffe2/test:distributed_gloo -- test_all_gather_coalesced

Differential Revision: D18123422

mrshenli commented on test/test_distributed.py:
```diff
@@ -1366,13 +1367,69 @@ def test_all_gather_cuda(self):
         self._test_all_gather_helper(group, group_id, rank)
 
+
+    def _test_all_gather_coalesced_helper(
```
This test case only covers 3D tensors. Let's add some tests to cover more corner cases.

1. input tensors are of different dimensions, e.g., `[torch.zeros(2, 2), torch.zeros(3, 3, 3), ...]` 
0. empty input tensors, e.g., `[torch.zeros([0])]` and `[torch.zeros([0]), torch.zeros([2, 2])]`

Let's put those in separate test cases so that when it fails, we know quickly what went wrong.

zappa- replied:
Done. 

BTW, passing only empty tensor breaks this, so we have to have at least non-empty one. I think that's a reasonable requirement. I've added it in description section.

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1158,6 +1160,63 @@ def all_gather(tensor_list,
+    # Flatten the input and create a list of flat outputs.
+    input_coalesced = torch.cat([t.flatten() for t in input_tensor_list])
+    output_coalesced = [
+        torch.empty(input_coalesced.numel()) for _ in output_tensor_lists]
```
This implicitly assumes size match. Let's make that explicit by adding size checks.

zappa- replied:
Done

mrshenli commented on test/test_distributed.py:
```diff
@@ -1366,13 +1367,69 @@ def test_all_gather_cuda(self):
+    def test_all_gather_coalesced_full_group(self):
+        group, group_id, rank = self._init_full_group_test()
+        self._test_all_gather_coalesced_helper(group, group_id, rank)
+
```
Let's add tests to make sure that when sizes input and output tensor lists do not match, we throw error with clear error message.

zappa- replied:
Done

mrshenli reviewed (commented):
Thanks for contributing! This looks good. I have the following general comments:

1. Please update this PR to remove debugging prints and comments
2. Lint CI failure is real, please address.
3. The main purpose of this PR is to add a minimum support for all_gather_coalesced and make sure we have good test coverage for that (not targeting perfs yet), so let's add more tests to cover corner cases.

zappa- commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1158,6 +1160,63 @@ def all_gather(tensor_list,
+    Returns:
+        Async work handle, if async_op is set to True.
+        None, if not async_op or if not part of the group
+
```
Done

zappa- commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1158,6 +1160,63 @@ def all_gather(tensor_list,
+    Arguments:
+        output_tensor_lists (list[list[Tensor]]): Output list. It should contain
+            correctly-sized tensors to be used for output of the collective.
+        input_tensor_list (list[Tensor]): Tensors to be broadcast from current process.
```
added comment, essentially similar to reduce_coalesced

zappa- commented on test/test_distributed.py:
```diff
@@ -1366,13 +1367,69 @@ def test_all_gather_cuda(self):
+        if group_id is not None:
+            for test_case_id in range(2, 5):
+                # Make sure we create tensors of incompatible sizes, e.g.
+                # [1x1], [2x2], ... to be sent in one batch
```
now building multidimensional tensors to address your other concern

zappa- commented:
addressed comments, please take another look.

mrshenli reviewed (approved):
LGTM! Thanks!

Commit:
adding python all_gather coalesced functionality and testing. (#28634)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/28634

caveat 1: this only works in sync mode.
caveat 2: this is going to go away and be replaced by c++ implementation

Test Plan: buck test caffe2/test:distributed_gloo -- test_all_gather_coalesced

Reviewed By: mrshenli

Differential Revision: D18123422

fbshipit-source-id: ee5c6dfa0ee9330f968948a9300d6372d20dfd6f
